mllib/src/main/scala/org/apache/spark/ml/recommendation/ALS.scala

```diff
@@ -395,13 +395,21 @@ class ALS(@Since("1.4.0") override val uid: String) extends Estimator[ALSModel]
       .map { row =>
         Rating(row.getInt(0), row.getInt(1), row.getFloat(2))
       }
+    val instr = Instrumentation.create(this, ratings)
+    instr.logParams(rank, numUserBlocks, numItemBlocks, implicitPrefs, alpha,
+                    userCol, itemCol, ratingCol, predictionCol, maxIter,
+                    regParam, nonnegative, checkpointInterval, seed)
     val (userFactors, itemFactors) = ALS.train(ratings, rank = $(rank),
       numUserBlocks = $(numUserBlocks), numItemBlocks = $(numItemBlocks),
       maxIter = $(maxIter), regParam = $(regParam), implicitPrefs = $(implicitPrefs),
       alpha = $(alpha), nonnegative = $(nonnegative),
       checkpointInterval = $(checkpointInterval), seed = $(seed))
     val userDF = userFactors.toDF("id", "features")
     val itemDF = itemFactors.toDF("id", "features")
+    val numUserFeatures = userDF.collect.size
+    val numItemFeatures = itemDF.collect.size
+    instr.logNumFeatures(numUserFeatures)
```

Comment:
In the instrumenter, this does `log(compact(render("numFeatures" -> num)))` so the message will be the same with different numbers, which is confusing as you then don't know which is user and item. I think we need a custom message here.